### PR TITLE
Bugfix for fitting compound models with mixed output units

### LINKED ISSUE #12320: Fitting Compound Models containing multiplied units fails
### Description

Creating a Compound model that multiplies the underlying output units returns expected units when evaluating, but fails during fitting because its trying to coerce all units to the data units and not accounting for the operators between the internal sub-models.

### Expected behavior

Fitting should succeed, with consistent behavior between units from calling evaluate and accepted during fitting.

### Actual behavior

Calling evaluate returns the expected units, but those same units as inputs while fitting raises errors.

### Steps to Reproduce

Below is a simplified (although non-physical) example:

```
import astropy.units as u
from astropy.modeling import models
from astropy.modeling.fitting import LevMarLSQFitter
import numpy as np

fitter = LevMarLSQFitter()

x = np.linspace(0,1,101) * u.s
y = np.linspace(5,10,101) * u.m * u.kg / u.s

m1 = models.Linear1D(slope=5*u.m/u.s/u.s, intercept=1.0*u.m/u.s
m2 = models.Linear1D(slope=0.0*u.kg/u.s, intercept=10.0*u.kg) 
c = m1 * m2
```

`c(x)` gives the expected units (m * kg / s), but `fitter(c, x, y)` fails.  Under-the-hood, the following is being called:

```
c._parameter_units_for_data_units({'x': x.unit}, {'y': y.unit})
```

which is evaluating to:

```
{'intercept_0': Unit("kg m / s"),
 'slope_0': Unit("kg m / s2"),
 'intercept_1': Unit("kg m / s"),
 'slope_1': Unit("kg m / s2")}
 ```
 
 In other words, `c._parameter_units_for_data_units` is attempting to make sure that each sub-model results in the passed output units, whereas it should make sure that the units _after applying operators_ result in the passed output units.  The following (in this case - obviously the general case is much more complicated) would make sense to be consistent with the forward model:
 
```
m1._parameter_units_for_data_units({'x': x.unit}, {'y': y.unit / m2(x).unit})
m2._parameter_units_for_data_units({'x': x.unit}, {'y': y.unit / m1(x).unit})
```

### System Details

macOS-10.15.7-x86_64-i386-64bit
Python 3.9.6 (default, Aug 18 2021, 12:38:10) 
[Clang 10.0.0 ]
Numpy 1.20.3
pyerfa 2.0.0
astropy 5.0.dev1030+g09720bb32
Scipy 1.7.1
Matplotlib 3.4.3

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

This PR fixes errors arising from fitting multiplication and division compound models where the sub models have different output units.

Fixes #12320

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by WilliamJamieson** [comment]
Tagging @kecnry, @nden, and @perrygreenfield for review.

**Review by kecnry — COMMENTED** [review]
I can confirm that this fixes the simplified test case as well as my original use case, with the exception of using a CompoundModel containing a BlackBody wrt wavelengths (frequencies are fixed with this PR).  The wavelength case is caused by a bug in the exposed `BlackBody.input_units` and is fixed with a combination of this PR and #12318.

**Review by perrygreenfield — COMMENTED** [review]
I must admit I struggled a bit to understand what was going on to make this work. I wonder if it would help to summarize the basic scheme somewhere in the code. Some of this difficulty is in divining the purpose of the methods from their names, particularly when there isn't a docstring or comments to explain. I think the addition of these would help a lot to make it easier to understand.

For example, by itself `without_units_for_data` doesn't really tell me what is supposed to go on (and the converse). Nor does _fix_parameter_units_for_data really explain what it is doing. Perhaps a really terse outline from what I understand would help. From what I see, the `without_units_for_data` recursively drills down the compound model tree, at each level sees what units for the output of its components are required by outputs of the level above. Eventually these hit the leaves of the expression, and it is in the parent model class that the without_units_for_data handles the parameter values and units to correspond to the inputs and outputs being asked of it. In the process of traversing the tree, information is embedded in the keywords to indicate what the implied units were, so that the reinstatement of the units can use this information in its recursive trip through the tree. Looks like to me that `_fix_parameter_units_for_data` is basically copy the parameter unit info from the original tree to the copy.

The code looks good to me with the mention that it doesn't seem like the test for `CompoundModel` in the parent model class needs to be done anymore since it should now never be called by `CompoundModel` since that subclass overrides this method.

One final comment is that perhaps in the documentation that dealing with units and fitting in complex compound models are potentially going to suffer significant performance penalties due to the massive evaluations triggered by requiring them at every level of the expression tree (plus forced generation of parameter names and such, which, IIRC, try to be bypassed if intermediate nodes never need to access them)

**Comment by WilliamJamieson** [comment]
@perrygreenfield see 1df0c22 for updates to the documentation to better explain the changes.

**Review by perrygreenfield — APPROVED** [review]
LGTM

**Comment by nden** [comment]
@pllim Is the failure in the 32-bit Ci job a known issue? Can't build numpy.

**Comment by pllim** [comment]
Re: 32-bit -- @saimn thought it was transient but maybe not...

**Comment by pllim** [comment]
RTD was terminated by the server for whatever reason. I just restarted it over there (the restart thingy is a new feature). 🤞

**Comment by pllim** [comment]
For future reference, please remember to add `backport-v5.0.x` label if you want it backported. Thanks.

@meeseeksdev backport to v5.0.x

**Comment by pllim** [comment]
@meeseeksdev backport to v5.0.x

File: astropy/modeling/tests/test_compound.py [hunk 1]
```diff
@@ -919,3 +920,149 @@ def test_compound_evaluate_fix_inputs_by_position():
...
+    fit = fitter(truth, x, y)
+
+    unfit_output = truth(x)
+    fit_output = fit(x)
+
+    assert unfit_output.unit == fit_output.unit == (u.kg * u.m / u.s)
+    assert_allclose(unfit_output, fit_output)
+
+    for name in truth.param_names:
+        assert getattr(truth, name) == getattr(fit, name)
```
Comment (dhomeier) [inline astropy/modeling/tests/test_compound.py:947]: This is asserting that the units are properly handled, but not that an actual fit is returned. If `fit` was the model fitted to `y`, `fit(x)` should return values close to `y`. Instead it returns results identical to the init model, and in fact the parameter check shows that the function parameters are unchanged from their starting values.

**Pull request merged**